plugin/arcconf/arcconf.py

```diff
@@ -0,0 +1,919 @@
+# License along with this library; If not, see <http://www.gnu.org/licenses/>.
+#
+# Author: Gris Ge <fge@redhat.com>
+#         Joe Handzik <joseph.t.handzik@hpe.com>
+#         Raghavendra Basvan <raghavendra.br@microsemi.com>
+
+import os
+import errno
+import re
+import json
+
+from pyudev import Context, Device, DeviceNotFoundError
+
+from lsm import (
+    IPlugin, Client, Capabilities, VERSION, LsmError, ErrorNumber, uri_parse,
+    System, Pool, size_human_2_size_bytes, search_property, Volume, Disk,
+    LocalDisk, Battery, int_div, JobStatus)
+
+from lsm.plugin.arcconf.utils import cmd_exec, ExecError
+
+_CONTEXT = Context()
+
+ZERO = 0
+COMPONENT = ['Adapter', 'Physical Device', 'Logical Device', 'Array', 'All', 'Unknown']
+
+#Drive states
+DRIVE_READY = 0
+DRIVE_ONLINE = 1
+DRIVE_RAW = 10
+DRIVE_STATE_UNKNOWN = 20
+
+#Controller states
+CONTROLLER_STATUS_WORKING = 0
+CONTROLLER_STATUS_FAILED = 1
+CONTROLLER_STATUS_INACCESSIBLE = 5
+CONTROLLER_STATUS_UNKNOWN = 6
+CONTROLLER_STATUS_LOCKED_UP = 9
+CONTROLLER_STATUS_SYS_PQI_DRIVER_CONFLICT = 10
+
+#RAID modes
+HW_RAID_EXPOSE_RAW_DEVICES = 0
+HBA_MODE = 2
+HW_RAID_HIDE_RAW_DEVICES = 3
+
+
+def _handle_errors(method):
+    def _wrapper(*args, **kwargs):
+        try:
+            return method(*args, **kwargs)
+        except LsmError:
+            raise
+        except KeyError as key_error:
+            raise LsmError(
+                ErrorNumber.PLUGIN_BUG,
+                "Expected key missing from arcconf output:%s" %
+                key_error)
+        except ExecError as exec_error:
+            if 'No controllers detected' in exec_error.stdout:
+                raise LsmError(
+                    ErrorNumber.NOT_FOUND_SYSTEM,
+                    "No Controllers deteceted by arcconf.")
+            else:
+                raise LsmError(ErrorNumber.PLUGIN_BUG, str(exec_error))
+        except Exception as common_error:
+            raise LsmError(
+                ErrorNumber.PLUGIN_BUG,
+                "Got unexpected error %s" % common_error)
+
+    return _wrapper
+
+
+def _sys_status_of(ctrl_status):
+    """
+    Based on data of "arcconf getconfig "
+    """
+    status_info = ''
+    status = System.STATUS_UNKNOWN
+    check_list = [
+        'Controller Status', 'Cache Status', 'Battery/Capacitor Status']
+    for key_name in check_list:
+        if key_name in ctrl_status:
+            if ctrl_status[key_name] == 'Not Configured':
+                status = System.STATUS_OK
+            elif ctrl_status[key_name] == 'OK':
+                status = System.STATUS_OK
+            elif ctrl_status[key_name] == 'Optimal':
+                status = System.STATUS_OK
+            else:
+                status = System.STATUS_OTHER
+            status_info += ' "%s"=[%s]' % (str(key_name),
+                                          str(ctrl_status[key_name]))
+
+    return status, status_info
+
+
+def _parse_arcconf_output(output, component):
```

Comment:
The design of arcconf is such that, getconfigJSON command requires a controller number as a parameter. eg: " arcconf getconfigJSON 1 " 
So the user should know before hand which controller information he/she needs. 
Hence, I had the implementation like :
a) issue one command first to know how many controllers are connected, eg: "arcconf list"  and b) issue getconfigJSON for each controller like "getconfigJSON <1>" , "getconfigJSON <2>", etc..  

I will try to avoid the parser code totally, and have a function that gives number of controllers found. Then use the getconfigJSON as required.